File: benchmark/nodejs/dfs.js
const n = 100;
const adj = [];
for (let i = 0; i < n; i++) {
    const row = [];
    for (let j = 0; j < n; j++) {
        row.push(0);
    }
    adj.push(row);
}
for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
        if ((i * 3 + j * 7) % 11 === 0) {
            if (i !== j) {
                adj[i][j] = 1;
            }
        }
    }
}
const start = Date.now();
const visited = [];
for (let i = 0; i < n; i++) {
    visited.push(false);
}
const stack = [0];
let count = 0;
while (true) {
    if (stack.length === 0) {
        break;
    }
    const node = stack.pop();
    if (!visited[node]) {
        visited[node] = true;
        count = count + 1;
        for (let neighbor = n - 1; neighbor >= 0; neighbor--) {
            if (adj[node][neighbor] === 1) {
                if (!visited[neighbor]) {
                    stack.push(neighbor);
                }
            }
        }
    }
}
const end = Date.now();
console.log(`Visited: ${count}`);
console.log(`Time: ${end - start}ms`);

File: benchmark/ruby/dfs.rb
n = 100
adj = []
n.times do |i|
    row = []
    n.times do |j|
        row << 0
    end
    adj << row
end
n.times do |i|
    n.times do |j|
        if (i * 3 + j * 7) % 11 == 0
            if i != j
                adj[i][j] = 1
            end
        end
    end
end
start = Time.now.to_f * 1000
visited = []
n.times do |i|
    visited << false
end
stack = [0]
count = 0
while true
    if stack.empty?
        break
    end
    node = stack.pop
    if not visited[node]
        visited[node] = true
        count = count + 1
        (n - 1).downto(0) do |neighbor|
            if adj[node][neighbor] == 1
                if not visited[neighbor]
                    stack.push(neighbor)
                end
            end
        end
    end
end
finish = Time.now.to_f * 1000
puts "Visited: #{count}"
puts "Time: #{(finish - start).round}ms"

